From d24457c19f18bac8ce8cf323834363e526f6ecf7 Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@gmail.com>
Date: Sun, 16 May 2021 10:52:18 -0700
Subject: [PATCH 101/196] Implement encoding CTxDestinations to XAddresses
 (#174)

Wraps XAddress::Encode in a helper function to directly create a
XAddress for Lotus from a CTxDestination. This type is used in
the wallet and RPC several places, and so support for this is
needed to connect XAddresses into the rest of the codebase.
---
 src/addresses/xaddress.cpp  | 28 ++++++++++++++++++++++
 src/addresses/xaddress.h    |  5 ++++
 src/test/xaddress_tests.cpp | 46 +++++++++++++++++++++++++++++++++++++
 3 files changed, 79 insertions(+)

diff --git a/src/addresses/xaddress.cpp b/src/addresses/xaddress.cpp
index 6e2d8afbe..92f495f5a 100644
--- a/src/addresses/xaddress.cpp
+++ b/src/addresses/xaddress.cpp
@@ -118,4 +118,32 @@ bool Parse(const CChainParams &params, const std::string &address,
     return ExtractDestination(scriptPubKey, retDestination);
 }
 
+class Encoder : public boost::static_visitor<std::string> {
+public:
+    explicit Encoder(const CChainParams &p) : params(p) {}
+
+    std::string operator()(const PKHash &id) const {
+        CScript scriptPubKey = CScript() << OP_DUP << OP_HASH160 << id
+                                         << OP_EQUALVERIFY << OP_CHECKSIG;
+        return Encode(Content(TOKEN_NAME, GetAddressNetworkByte(params),
+                              SCRIPT_PUB_KEY, ToByteVector(scriptPubKey)));
+    }
+
+    std::string operator()(const ScriptHash &id) const {
+        CScript scriptPubKey = CScript() << OP_HASH160 << id << OP_EQUAL;
+        return Encode(Content(TOKEN_NAME, GetAddressNetworkByte(params),
+                              SCRIPT_PUB_KEY, ToByteVector(scriptPubKey)));
+    }
+
+    std::string operator()(const CNoDestination &) const { return ""; }
+
+private:
+    const CChainParams &params;
+};
+
+std::string EncodeDestination(const CChainParams &params,
+                              const CTxDestination &dst) {
+    return boost::apply_visitor(Encoder(params), dst);
+}
+
 } // namespace XAddress
diff --git a/src/addresses/xaddress.h b/src/addresses/xaddress.h
index 6933588b0..4d53c3270 100644
--- a/src/addresses/xaddress.h
+++ b/src/addresses/xaddress.h
@@ -63,6 +63,11 @@ DecodeError Decode(const std::string &address, Content &parsedOutput);
  */
 bool Parse(const CChainParams &params, const std::string &address,
            CTxDestination &retDestination);
+/**
+ * Encodes a CTxDestination to an XAddress. Returns the empty string on failure.
+ */
+std::string EncodeDestination(const CChainParams &params,
+                              const CTxDestination &dst);
 } // namespace XAddress
 
 #endif // BITCOIN_XADDR_H
diff --git a/src/test/xaddress_tests.cpp b/src/test/xaddress_tests.cpp
index a2b8c4368..738959a2f 100644
--- a/src/test/xaddress_tests.cpp
+++ b/src/test/xaddress_tests.cpp
@@ -194,4 +194,50 @@ BOOST_AUTO_TEST_CASE(parse_p2pkh_works) {
     }
 }
 
+BOOST_AUTO_TEST_CASE(encode_destination_works) {
+    // First create an instance of an engine.
+    std::random_device rd;
+    auto gen = [&rd]() { return uint8_t(rd()); };
+    GlobalConfig config;
+    std::vector<uint8_t> fakeKeyBytes(32);
+    std::generate(begin(fakeKeyBytes), end(fakeKeyBytes), gen);
+    const auto fakeKeyId = Hash160(fakeKeyBytes);
+    // Assert that encode works for PKHash.
+    {
+        CScript scriptPubKey = CScript() << OP_DUP << OP_HASH160
+                                         << ToByteVector(fakeKeyId)
+                                         << OP_EQUALVERIFY << OP_CHECKSIG;
+        const std::string encodedAddress = XAddress::Encode(XAddress::Content(
+            XAddress::TOKEN_NAME, XAddress::MAINNET, XAddress::SCRIPT_PUB_KEY,
+            ToByteVector(scriptPubKey)));
+        BOOST_CHECK_EQUAL(
+            XAddress::EncodeDestination(config.GetChainParams(),
+                                        PKHash(CKeyID(fakeKeyId))),
+            encodedAddress);
+    }
+
+    // Assert that encode works for ScriptHash.
+    {
+        CScript redeemScript = CScript() << OP_TRUE;
+        CScript scriptPubKey =
+            CScript() << OP_HASH160
+                      << ToByteVector(Hash160(ToByteVector(redeemScript)))
+                      << OP_EQUAL;
+
+        const std::string encodedAddress = XAddress::Encode(XAddress::Content(
+            XAddress::TOKEN_NAME, XAddress::MAINNET, XAddress::SCRIPT_PUB_KEY,
+            ToByteVector(scriptPubKey)));
+        BOOST_CHECK_EQUAL(XAddress::EncodeDestination(config.GetChainParams(),
+                                                      ScriptHash(redeemScript)),
+                          encodedAddress);
+    }
+
+    // Assert null address for CNoDestination.
+    {
+        BOOST_CHECK_EQUAL(XAddress::EncodeDestination(config.GetChainParams(),
+                                                      CNoDestination()),
+                          "");
+    }
+}
+
 BOOST_AUTO_TEST_SUITE_END()
-- 
2.39.5 (Apple Git-154)

